### ISSUE TITLE
TabNavigator tab button tap causes impossibly slow transition when remote debugging is off

### ISSUE THREAD

**Issue by uriva** [react-community/react-navigation#1454]
Using react-native 0.43 or 0.44.0, I can't get the TabNavigator to work properly.

Switching tabs via the buttons is super slow, but via swipe motion is fast.

For some reason, when remote debugging on chrome, everything is fast, but when I turn it off, it's slow again (counter-intuitive).

There are no logging whatsoever and JS dev mode is off.

This behaviour happens both on iOS and android.

Since this happens only when js remote debugging is off, the release apk is slow.

**Comment by AlanFoster** [comment]
@uriva If you turn off animations, is there improved performance increase?

**Comment by uriva** [comment]
I tried now - makes no difference.

**Comment by AlanFoster** [comment]
@uriva Do you have more details; i.e. are you running on an emulator, real phone? what version of react navigation? Do you have custom onPress listener code that's slow when pressing the navigation? How expensive is your re-render? Do you have an example project we can test out?

**Comment by uriva** [comment]
Some code:

```
import React, { Component } from 'react';
import { Container, Icon, StyleProvider } from 'native-base';
import { connect } from 'react-redux';
import { Platform } from 'react-native';
import { TabNavigator } from 'react-navigation';
import getTheme from '../native-base-theme/components';
import { iconFamilyTheme } from '../native-base-theme/variables/platform';
import { NavigationActions } from 'react-navigation';

import Items from './items';
import Header from './header';

import categories from '../common/categories';

const isIos = Platform.OS === 'ios';

const categorizedItems = category => props => (
  <Items category={category} {...props} />
);

const categoriesRouteConfigs = Object.keys(
  categories
).reduce((res, category) => {
  const categoryObject = categories[category];
  res[category] = {
    screen: categorizedItems(category),
    navigationOptions: {
      tabBarIcon: (focused, tintColor) => (
        <StyleProvider
          style={getTheme(iconFamilyTheme(categoryObject.iconFamily))}
        >
          <Icon name={categoryObject.icon} />
        </StyleProvider>
      )
    }
  };

  return res;
}, {});

const HomepageNavigator = TabNavigator(categoriesRouteConfigs, {
  initialRouteName: 'firstTab',
  lazy: false,
  animationEnabled: false,
  tabBarOptions: {
    showIcon: true,
    showLabel: isIos,
    style: isIos
      ? {}
      : {
          backgroundColor: '#3b5a94'
        }
  }
});

class Homepage extends Component {
  componentWillUpdate(nextProps) {
    const { notificationCategory } = nextProps;
    const action = NavigationActions.navigate({
      routeName: notificationCategory
    });
    this.refs.innerNavigator.dispatch(action);
  }

  render() {
    const { changeTab, navigation } = this.props;

    return (
      <Container>
        <Header hasTabs title="some title" noBack navigation={navigation} />
        <HomepageNavigator
          ref="innerNavigator"
          screenProps={{ rootNavigation: navigation }}
          onNavigationStateChange={(prevState, newState) => {
            const { routes, index } = newState;
            const { routeName } = routes[index];
            if (routeName in categories) {
              changeTab(routeName);
            }
          }}
        />
      </Container>
    );
  }
}

const mapStateToProps = state => {
  return {
    notificationCategory: state.notificationCategory
  };
};

const mapDispatchToProps = dispatch => ({
  changeTab: tabId =>
    dispatch({
      type: 'CHANGE_TAB',
      tabId
    })
});

export default connect(mapStateToProps, mapDispatchToProps)(Homepage);
```

**Comment by uriva** [comment]
I also neglected to mention that when swiping, the tab itself moves immediately, but the tab icons animation lags behind a few seconds. When tapping on the tab icons everything is delayed a few seconds.

All this happens only when debug js remotely is off.

**Comment by uriva** [comment]
I think it might have something to do with the yellow line under the tab icons, when switching with swipe motion, it lags a few seconds behind, but if you swipe and not release, it will eventually come back to life and then the transition is fast.

Can I somehow disable that line animation?

**Comment by uriva** [comment]
react navigation version is 1.0.0-beta.9
the described issue happens only on real device, on genymotion it is better

**Comment by uriva** [comment]
@AlanFoster I don't think it is because of a heavy re-render, because this still occurs when the tab contents is an empty view.

**Comment by koenpunt** [comment]
Can it be that you have enabled slow animations in the simulator?

**Comment by andfs** [comment]
@uriva  have you found a solution?

**Comment by uriva** [comment]
Various optimizations helped a bit, specifically getting rid of the native base svg icons (something doesn't work well there) and limiting as much as possible the re-read.

However the end result clearly indicates a library problem or something wrong with react native itself, and the speed is reasonable but far from ideal.

**Comment by nemanja-stanarevic** [comment]
@uriva any chance you can go over getting rid of native base svg icons? We are using custom png icons and still see performance issues with tab navigator.

**Comment by zsomborgellerfi** [comment]
+1

**Comment by andfs** [comment]
The new version 1.0.0-beta.10 is slower then never in android.
My tabs are empty and is extremely slow. I'll be workin on beta9 and wait for a solution untill I release the app :/

**Comment by arnaubennassar** [comment]
Same here, super slow animations when remote debugging is off. On iphone emulator and android rela device, emulator, and release in device. Using a stack navigator neasted into a tab navigator. (The tab navigator work smoothly, but the stack is super slow)

**Comment by arnaubennassar** [comment]
I have removed some console.log and now it's usable : D

**Comment by WillyRamirez** [comment]
@arnaubennassar  did you remove console.logs in your app folder alone? I removed all of my own, however a lot of them remain in the node_modules folder

**Comment by WillyRamirez** [comment]
#608

**Comment by arnaubennassar** [comment]
Only in my folder. I would say that as long as they dont get executed its ok. But if for some reason some modules are doing it too much or too heavy, that may be the cause.

**Comment by devarthurribeiro** [comment]
+1

**Comment by amed-kino** [comment]
It seems that there is no efficient diagnose for the issue. Since I removed most of the libraries trying to catch the lazy part, it did not help performance enhance and I still don't know what is wrong with `^1.0.0-beta.11`

**Comment by amed-kino** [comment]
**Suloution**:
The App used [Sentry](https://sentry.io) to track errors, `Stacktrace` was getting effect by the new version. 

https://github.com/getsentry/react-native-sentry/issues/158

**Comment by devarthurribeiro** [comment]
+1

**Comment by l1br3** [comment]
I've got the exact opposite issue as OP with beta 11, action dispatching is over slow when remote debug is ON on android emulator API 26.
StackNavigators nested in TabNavigator

**Comment by kelset** [comment]
Closing this issue in favor of https://github.com/react-community/react-navigation/issues/739 (and also look here for [more info](https://github.com/react-community/react-navigation/issues/608#issuecomment-328635042) related to performances)

**Comment by rafaelgrilli92** [comment]
Here is my workaround. I wrapped the tab header content inside a **TouchableWihoutFeedback** component and then the "button" style I overlapped the whole content. When the user clicks on the tab, he is actually pressing the button and then on the "onPress" method, I use the tab reference to call the method "goToPage" passing which index tab the user is pressing and it worked perfectly for me.

```
<TabHeading style={{
	flex: 1,
	flexDirection: 'column',
	borderTopWidth: .5,
	borderTopColor: bottomMenu.borderTopColor,
	backgroundColor: bottomMenu.backgroundColor
}}>
	<TouchableWithoutFeedback 
		**style={{
			height: 50,
			alignSelf: 'stretch',
			paddingVertical: 5
		}}** 
		onPress={() => this.tabs.goToPage(tabIndex)}
	>
		// Tab header content. Can be texts, icons, whatever
	</TouchableWithoutFeedback>
</TabHeading>
```

However, when I swipe between tabs, it takes around 1 second to actually change the tabs. So, the content changes instantly, but the tabs layout takes 1 second to change. Even with the simple example.

RN: 0.51
NB: 2.3.9

**Comment by ppozniak** [comment]
I see #739 is marked as resolved but don't see resolution except for console.logs, I've tried every suggested solutions from this issue and from 
For me this problem still occurs. Someone in the comments mentioned this may be connected to NativeBase. Any ways of fixing this without dropping out the whole library?

Different comments suggest that this is connected to redux, maybe it's worth to drop deeper into this?

And no this is not about logs. I've got rid of redux logger and have 0 console.logs. Also with or without lazy prop (doesn't have any effect)
*This also occurs on production*

So basically moving between screens by swiping works really smooth, however pressing a tab creates a 
big delay, also active state on the tabs updates really slow (in this case even when swiping).

Now I'm using `FooterTab` from NativeBase and the delay is a little bit slower, however this was also happening with default tabBarComponent of react navigation.

So is this really NativeBase problem? Should I create an issue over

**Comment by satya164** [comment]
> however pressing a tab creates a big delay

This is a sign of unnecessary re-renders. Tab switching by tapping should be super fast unless too much is going on. By default, React Navigation should optimize this. But if this is happening in your app, that means there's something causing re-renders. A common source of this is `screenProps`. You should profile your app to see what's taking time.

**Comment by sandywk** [comment]
Is there a solution now?
react-navigation 1.0.0-beta.21
Can the updated version solve this problem?

**Comment by gilons** [comment]
I'm posting this because i noticed it's a closed issue but people are still having problem unsolved.
i encountered the same problem days ago . I investigated and i found that the component to be rendered by the tab was taking too much time to render so the tab navigator has to wait until the component is rendered .
I had a componentDidMount method where i was just setting a state . I tried conditional rendering, it did not solve the issue . 
The conditional rendering was as follows
`
  render() {
    return this.state.isLoading?<ActivityIndicator/>:<MyHeavyComponent/>
  }`
 and my componentDidMount was follows 
`componentDidMount() {
      this.setState({
        isLoading: false,
        Events: cardListData
      });
  }`

My intention was to make the tab to navigate fast and first render the <activityIndicator/> component and then, when the component has finished mounting render <MyHeavyComponent/> .  but it did not solve my problem .
So the solution that worked for me was to used setTimeout function to slow down the `this.setState` in the ComponentDidMount method as follows 

`componentDidMount() {
    setTimeout(() => {
      this.setState({
        loadingEvents: false,
        Events: cardListData
      });
    }, 100)

  }`
   so that the tab should have time to render the <ActivityIndicator/> . this fasten the tab navigation .

**Comment by gilons** [comment]
You can adjust the timeout duration depending on your user case

**Comment by NelsoMart** [comment]
gilons, but where in your code do you use this? I do not understand.
How do you use this:
componentDidMount() { setTimeout(() => { this.setState({ isloading: false, }); }, 100) }

**Comment by gilons** [comment]
That is the componentDidMount of the component rendered by the tab to which you are navigating

**Comment by NelsoMart** [comment]
Ok, it's just that I don't understand where I have to place this component to make it work!

componentDidMount() { setTimeout(() => { this.setState({ isloading: false, }); }, 100) }

**Comment by gilons** [comment]
In the component class definition.

**Comment by gilons** [comment]
In the component class definition.

**Comment by gilons** [comment]
In the component class definition.

**Comment by gilons** [comment]
In the component class definition.

**Comment by NelsoMart** [comment]
Ok. gilons, Thankyou.

**Comment by gilons** [comment]
Ok good!
